control_protocol.md

```diff
@@ -0,0 +1,293 @@
+# Control protocol
+
+The control protocol transmits messages via its {ref}`transport-layer` from one Component to another one.
+The {ref]}`message-layer` is the common language to understand commands, thus creating a remote procedure call.
+
+
+(transport-layer)=
+## Transport layer
+
+The transport layer ensures, that a message arrives its destination.
+
+
+### Sockets and Connections
+
+We use [zmq](https://zeromq.org/) sockets for our communication. For more details see the [zmq guide](https://zguide.zeromq.org/) or [zmq API](http://api.zeromq.org/)
+
+
+#### Definitions
+
+Zmq messages consist in a series of frames, each is a byte sequence.
+We indicate the separation between frames with `|`.
+An empty frame is indicated with two frame separators `||`, even at the beginning or end of a message.
+For example `||Second frame|Third frame||Fifth frame` consists of 5 frames.
+The first and fourth frames are empty frames.
+
+
+#### Configuration
+
+Each {ref}`Coordinator <components.md#coordinator>` shall offer one ROUTER socket, bound to a host name (or IP address or any address of a computer with "*") and port.
+
+{ref}`Components <components.md#components>` shall have one DEALER socket connecting to one Coordinator's ROUTER socket.
+
+Coordinators shall have one DEALER socket per other Coordinator in the Network.
+This DEALER socket shall connect to the other Coordinator's ROUTER socket.
+
+Messages must be sent to a Coordinator's ROUTER socket.
+
+
+(router-sockets)=
+#### Particularities of ROUTER sockets
+
+A small introduction to ROUTER sockets, for more details see [zmq guide chapter 3](https://zguide.zeromq.org/docs/chapter3/#Exploring-ROUTER-Sockets).
+
+A router socket assigns a random _identity_ to each connecting peer.
+If, for example, two Components `CA`, `CB` connect to the ROUTER sockets, the socket assigns the identities, which we call `IA`, `IB` (they can be any bytes sequence).
+
+Whenever a message is sent to the ROUTER socket from a peer, it reads as a first frame that identity.
+For example if `CA` sends the message `Request A`, the ROUTER socket will read `IA|Request A`.
+That way, you can return an answer to exactly the original peer and not, for example `CB`.
+If you call the ROUTER's send command with `IA|Reply A`, the socket will send `Reply A` to the peer, whose connection is `IA`, in this case `CA`.
+
+The following diagram shows this example communication with two Components:
+:::{mermaid}
+graph TD
```

Comment:
Conceptually, a ROUTER socket has one "send" and one "recv" node (method), so I find the display here a bit confusing. 
Maybe it would be better to split into two separate graphs, first showing the flow for CA, then for CB? But then, they will be ~identical, so we don't learn that much.
Maybe this would be better as a sequence diagram with participants CA, CB, send, recv, Coordinator, and you show the flow how messages exchanged with CA and CB flow through send/recv differently? Do you know what I mean? If not I can sketch it.

Reply:
I think I know, what you mean:
CA to Co1 recv. Co1 send to CO2 recv etc.?

Reply:
Looking at the diagram, I understand you.
Is that, what you intend:
~~~mermaid
sequenceDiagram
    participant Code
    participant ROUTER as ROUTER socket
    Note over Code, ROUTER: Coordinator
    CA ->> ROUTER: "Request A"
    ROUTER ->> Code: "IA|Request A"
    Code ->> ROUTER: "IA|Reply A"
    ROUTER ->> CA: "Reply A"
    CB ->> ROUTER: "Request B"
    ROUTER ->> Code: "IB|Request B"
    Code ->> ROUTER: "IB|Reply B"
    ROUTER ->> CB: "Reply B"
~~~

Reply:
Exactly!

Reply:
I am not sure about calling it "Coordinator internals", since already the ROUTER socket is an (internal) part of the Coordinator, although it does make one connection to the outside. I am not perfectly sure about the elegance of calling it "Code", but I think it is clearly understandable, and that is the main goal of it, I think, so for me it is fine.

Reply:
Followong that reasoning, the Router Socket is also "Code", though, isn't it?

Reply:
What about "Message handling"?

Reply:
Sounds good to me!

Reply:
sounds good to me too. 

For reference and explanation, I made the distinction here between the direct _application "Code"_, controlled directly by somebody implementing this protocol, and the _zmq socket_, which although undoubtedly "lives" in some sense in "code", (not the application code implementing the protocol) is not controlled/touched in any way by a user who implements the protocol, treated as a black box outside. Into this black box, the "Code" stuffs whatever it is written to stuff, and gets out whatever the zmq socket will provide.

Reply:
(btw Benjamin, have you seen our Matrix room? Want to joint? Could be a good way to get a video call set up one weekend https://matrix.to/#/#pymeasure-ecp:matrix.org)

Reply:
thanks for the reminder!